File: backend/internal/model/media.go
package model

// Media represents a logical media item (movie or episode)
type Media struct {
	ID              int64        `json:"id"`
	LibraryID       int64        `json:"library_id"`
	MediaType       string       `json:"media_type"` // "movie" | "episode"
	Title           string       `json:"title"`
	SortTitle       string       `json:"sort_title"`
	TmdbID          *int64       `json:"tmdb_id,omitempty"`
	ImdbID          *string      `json:"imdb_id,omitempty"`
	TvdbID          *int64       `json:"tvdb_id,omitempty"`
	Overview        string       `json:"overview"`
	Tagline         string       `json:"tagline"`
	ReleaseDate     string       `json:"release_date"` // YYYY-MM-DD
	Rating          float64      `json:"rating"`
	IMDbRating      float64      `json:"imdb_rating"`
	RTScore         int          `json:"rt_score"`
	MetacriticScore int          `json:"metacritic_score"`
	PosterPath      PosterPath   `json:"-"`
	BackdropPath    BackdropPath `json:"-"`
	LogoPath        LogoPath     `json:"-"`
	ThumbPath       BackdropPath `json:"-"`

	Poster   *ImageResource `json:"poster,omitempty"`
	Backdrop *ImageResource `json:"backdrop,omitempty"`
	Logo     *ImageResource `json:"logo,omitempty"`
	Thumb    *ImageResource `json:"thumb,omitempty"`

	MetadataLocked bool   `json:"metadata_locked"`
	CreatedAt      string `json:"created_at"`
	UpdatedAt      string `json:"updated_at"`
}

// MetadataEditRequest represents a partial metadata edit for media.
// Pointer fields: nil = don't change. Slice fields: nil = don't change, empty = clear all.
type MetadataEditRequest struct {
	Title          *string       `json:"title"`
	SortTitle      *string       `json:"sort_title"`
	Overview       *string       `json:"overview"`
	Tagline        *string       `json:"tagline"`
	ReleaseDate    *string       `json:"release_date"`
	Rating         *float64      `json:"rating"`
	Genres         []string      `json:"genres"`
	Credits        []CreditInput `json:"credits"`
	SaveNFO        bool          `json:"save_nfo"`
	MetadataLocked *bool         `json:"metadata_locked"`
}

// CreditInput represents a credit entry in a metadata edit request.
type CreditInput struct {
	PersonName string `json:"person_name"`
	Character  string `json:"character,omitempty"`
	Role       string `json:"role"` // "cast" | "director" | "writer"
	Order      int    `json:"order"`
}

// MediaFile represents a physical video file on disk
type MediaFile struct {
	ID             int64   `json:"id"`
	MediaID        int64   `json:"media_id"`
	FilePath       string  `json:"file_path"`
	FileSize       int64   `json:"file_size"`
	Duration       float64 `json:"duration"`
	Width          int     `json:"width"`
	Height         int     `json:"height"`
	VideoCodec     string  `json:"video_codec"`
	VideoProfile   string  `json:"video_profile,omitempty"`
	VideoLevel     int     `json:"video_level,omitempty"`
	VideoFPS       float64 `json:"video_fps,omitempty"`
	AudioCodec     string  `json:"audio_codec"`
	Container      string  `json:"container"`
	Bitrate        int     `json:"bitrate"`
	IsHDR          bool    `json:"is_hdr,omitempty"`
	DVProfile      int     `json:"dv_profile,omitempty"`
	ColorTransfer  string  `json:"color_transfer,omitempty"`
	ColorPrimaries string  `json:"color_primaries,omitempty"`
	Fingerprint    string  `json:"fingerprint"` // "{file_size}:{xxhash64_first_64KB}"
	IsPrimary      bool    `json:"is_primary"`
	AddedAt        string  `json:"added_at"`
	LastVerifiedAt *string `json:"last_verified_at,omitempty"`
}

// MediaWithFiles combines media with its files
type MediaWithFiles struct {
	Media Media       `json:"media"`
	Files []MediaFile `json:"files"`

	// Episode-only fields (populated when media_type == "episode")
	SeriesID      int64 `json:"series_id,omitempty"`
	SeasonID      int64 `json:"season_id,omitempty"`
	EpisodeNumber int   `json:"episode_number,omitempty"`
	SeasonNumber  int   `json:"season_number,omitempty"`
}

// MediaListItem represents a media item for list views with genres
type MediaListItem struct {
	ID          int64          `json:"id"`
	Title       string         `json:"title"`
	SortTitle   string         `json:"sort_title"`
	PosterPath  PosterPath     `json:"-"`
	Poster      *ImageResource `json:"poster,omitempty"`
	MediaType   string         `json:"media_type"`
	Genres      []string       `json:"genres"`
	SeriesID    int64          `json:"series_id,omitempty"`
	ReleaseDate string         `json:"release_date,omitempty"`
	Rating      float64        `json:"rating,omitempty"`
	Overview    string         `json:"overview,omitempty"`
	// Progress (from user_data, populated when UserID filter is set)
	Position  *float64 `json:"position,omitempty"`
	Duration  *float64 `json:"duration,omitempty"`
	Completed *bool    `json:"completed,omitempty"`
}

// MediaListFilter represents filter parameters for media list queries
type MediaListFilter struct {
	UserID    int64 // populated from auth context for progress join
	LibraryID int64
	MediaType string // "movie" | "episode" | ""
	Search    string // LIKE on title + sort_title
	Genre     string // exact match genre name
	Year      string // 4-digit year string
	Sort      string // "newest" | "oldest" | "rating" | "title"
	Limit     int
	Offset    int
}

File: webapp/src/types/media.ts
// Media, Library, Genre, Browse, Search types

export interface Library {
  id: number
  name: string
  type: string
  paths: string[]
  created_at: string
}

export interface CreateLibraryRequest {
  name: string
  type: string
  paths: string[]
}

export interface Media {
  id: number
  library_id: number
  media_type: 'movie' | 'episode'
  title: string
  sort_title: string
  tmdb_id?: number
  imdb_id?: string
  tvdb_id?: number
  overview: string
  tagline: string
  release_date: string
  rating: number
  imdb_rating: number
  rt_score: number
  metacritic_score: number
  poster_path: string
  backdrop_path: string
  logo_path: string
  thumb_path: string
  metadata_locked: boolean
  duration?: number // from media_files join
  series_id?: number // for episodes
  season_id?: number // for episodes
  created_at: string
  updated_at: string
}

export interface MediaFile {
  id: number
  media_id: number
  file_path: string
  file_size: number
  duration: number
  width: number
  height: number
  video_codec: string
  audio_codec: string
  container: string
  bitrate: number
  fingerprint: string
  is_primary: boolean
  added_at: string
  last_verified_at?: string
}

export interface MediaWithFiles {
  media: Media
  files: MediaFile[]
  // Episode-only fields
  series_id?: number
  season_id?: number
  episode_number?: number
  season_number?: number
}

export interface MediaListItem {
  id: number
  title: string
  sort_title: string
  poster_path: string
  media_type: 'movie' | 'episode'
  type?: 'movie' | 'series' // frontend-friendly alias
  genres: string[]
  release_date?: string
  rating?: number
  overview?: string
  series_id?: number
}

export interface MediaListParams {
  library_id?: number
  type?: 'movie' | 'episode'
  // Filter params (Plan M)
  search?: string
  genre?: string
  year?: string
  sort?: 'newest' | 'oldest' | 'rating' | 'title'
  limit?: number
  offset?: number
}

export interface MetadataEditRequest {
  title?: string
  sort_title?: string
  overview?: string
  tagline?: string
  release_date?: string
  rating?: number
  genres?: string[]
  credits?: CreditInput[]
  save_nfo?: boolean
  metadata_locked?: boolean
}

export interface CreditInput {
  person_name: string
  character?: string
  role: 'cast' | 'director' | 'writer'
  order: number
}

export interface CreditWithPerson {
  credit: {
    id: number
    person_id: number
    character: string
    role: string
    display_order: number
  }
  person: {
    id: number
    name: string
    profile_path: string
  }
}

export interface Genre {
  id: number
  name: string
  tmdb_id?: number
}

export interface BrowseFolderItem {
  name: string
  path: string
  media_count?: number
  poster?: string
}

export interface BrowseResult {
  library_id: number
  path: string
  parent: string
  folders: BrowseFolderItem[]
  media: MediaListItem[]
}

export interface SearchResult {
  movies: MediaListItem[]
  series: SeriesListItem[]
}

// SeriesListItem is defined in series.ts — import for SearchResult
import type { SeriesListItem } from './series'
export type { SeriesListItem } from './series'
